### PR TITLE
etcd3 store: provide compactor util

### PR DESCRIPTION
**Pull request by hongchaodeng**
What's this PR?
- Provides a util to compact keys in etcd.

Reason:
We want to save the most recent 10 minutes event history. It should be more than enough for slow watchers. It is not number based, so it can tolerate event bursts too. We do not want to save longer since the current storage API cannot take advantage of the multi-version key yet. We might keep a longer history in the future.

### CODE REVIEW

**Comment by hongchaodeng** [comment]
ref: #22448 #24022

File: pkg/storage/etcd3/util.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+			return
+		}
+
+		getResp, err := client.Get(ctx, "/")
+		if err != nil {
+			glog.Errorf("compactor: Get failed: %v", err)
+			continue
+		}
+		rev := getResp.Header.Revision
+		if err := client.Compact(ctx, rev); err != nil {
```
Comment (xiang90) [inline pkg/storage/etcd3/util.go]: shouldn't we record the current one as prev and compact the prev when we get the next rev after some interval?

Comment (xiang90) [inline pkg/storage/etcd3/util.go]: basically we need to move line 38 before line 32

**Comment by hongchaodeng** [comment]
Fixed. Sorry for the mistake :p

**Comment by xiang90** [comment]
LGTM

File: pkg/storage/etcd3/util.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+func compactor(ctx context.Context, client *clientv3.Client, interval time.Duration) {
+	var curRev int64
+	for {
+		select {
+		case <-time.After(interval):
+		case <-ctx.Done():
+			return
+		}
+
+		getResp, err := client.Get(ctx, "/")
```
Comment (xiang90) [inline pkg/storage/etcd3/util.go]: probably just resp instead of getResp. there should not be any confusion.

File: pkg/storage/etcd3/util.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+import (
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/golang/glog"
+	"golang.org/x/net/context"
+)
+
+// compactor periodically compacts keys in etcd. After compaction, old versions of
```
Comment (xiang90) [inline pkg/storage/etcd3/util.go]: compact historical version of keys

File: pkg/storage/etcd3/util.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+// keys will be gone. Any API call for the old version will return error.
```
Comment (xiang90) [inline pkg/storage/etcd3/util.go]: old versions of keys set before the given interval will be gone.

File: pkg/storage/etcd3/util.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+// interval: the interval between each compaction. The first compaction happens after "interval".
+func compactor(ctx context.Context, client *clientv3.Client, interval time.Duration) {
```
Comment (xiang90) [inline pkg/storage/etcd3/util.go]: startCompactor? or runCompactor?

Comment (hongchaodeng) [inline pkg/storage/etcd3/util.go]: It looked better like this:
```
go compactor(...)
```
Very personal preference...

**Comment by hongchaodeng** [comment]
Addressed comments

File: pkg/storage/etcd3/util.go [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+	"github.com/coreos/etcd/clientv3"
+	"github.com/golang/glog"
+	"golang.org/x/net/context"
+)
+
+// compactor periodically compacts historical versions of keys in etcd.
+// After compaction, old versions of keys set before given interval will be gone.
+// Any API call for the old versions of keys will return error.
+// interval: the interval between each compaction. The first compaction happens after "interval".
+func compactor(ctx context.Context, client *clientv3.Client, interval time.Duration) {
```
Comment (timothysc) [inline pkg/storage/etcd3/util.go]: Do you intend to expose through the interface?

Comment (xiang90) [inline pkg/storage/etcd3/util.go]: my thought is no. we should not expose this directly. but we might want to make the interval configurable when creating the etcd3 storage implementation.

Comment (hongchaodeng) [inline pkg/storage/etcd3/util.go]: No.

I'm expecting to run a goroutine in a New function in etcd3 package. Like
```Go
package etcd3

func New(...) *store {
  go compactor(ctx, client, cfg.CompactionInterval)
  return newStore(...)
}

```

File: pkg/storage/etcd3/util.go [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/golang/glog"
+	"golang.org/x/net/context"
+)
+
+// compactor periodically compacts historical versions of keys in etcd.
+// After compaction, old versions of keys set before given interval will be gone.
+// Any API call for the old versions of keys will return error.
```
Comment (timothysc) [inline pkg/storage/etcd3/util.go]: This seems like a destructive operation, when and why would we do this?

File: pkg/storage/etcd3/util.go [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+		curRev = resp.Header.Revision
+		if oldRev == 0 {
+			continue
+		}
+		err = client.Compact(ctx, oldRev)
+		if err != nil {
+			glog.Errorf("compactor: Compact failed: %v", err)
+			continue
+		}
+		glog.V(4).Infof("compactor: Compacted rev %d", oldRev)
```
Comment (timothysc) [inline pkg/storage/etcd3/util.go]: I would regular Infof b/c I'm guessing the interval level is in O(minutes)

File: pkg/storage/etcd3/util.go (hunk 7 above)
Comment (xiang90) [inline pkg/storage/etcd3/util.go]: at some point, we need to compact the history. we should not allow the history to grow unboundedly. Right now, no consumer should even expect to consume historical data before 1000 or say 10000 versions due to etcd2's implementation. Also the storage API does not expose this details. We can pace the compaction in storage itself as long as the API stays the same. Once we do want to expose historical versions of keys, we might change the default behavior.

File: pkg/storage/etcd3/util.go (hunk 6 above)
Comment (hongchaodeng) [inline pkg/storage/etcd3/util.go]: Sorry. I was wrong. We should run in init way. Once we have known information about all cluster, we should do
```Go
func InitCompactors(...){
  c1 := createClient(cfg1)
  go compactor(c1, ...)
  c2 := createClient(cfg2)
  go compactor(c2, ...)
}
```

Maybe we should expose InitCompactors() then? I need to think about this further. But we could still review the code. This is the implementation details. We can deal with public APIs later.

Comment (timothysc) [inline pkg/storage/etcd3/util.go]: In a multiple api-server configuration how exactly would multiple compactors work?

Comment (timothysc) [inline pkg/storage/etcd3/util.go]: could we just name this compator.go ?  There are literally dozens of util.go files in the repository.

**Comment by hongchaodeng** [comment]
Addressed comments and do some refactoring.
@timothysc PTAL

**Comment by timothysc** [comment]
LGTM.

**Comment by hongchaodeng** [comment]
@timothysc @smarterclayton 
Can you add 'ok-to-merge'?

**Pull request merged**